### PR TITLE
Add agg cache

### PR DESCRIPTION
**Pull request by czpmango**
> 

Add aggregateExpression's runtime cache.Just Follow [https://github.com/vesoft-inc/nebula-common/pull/407](https://github.com/vesoft-inc/nebula-common/pull/407) to achieve this.

---

Benchmark:
Before this PR:
c1

After this PR:
c2

### CODE REVIEW

File: src/common/expression/AggregateExpression.cpp [hunk 1]
```diff
@@ -43,15 +47,12 @@ const Value& AggregateExpression::eval(ExpressionContext& ctx) {
         uniques->values.emplace(val);
     }
 
-    apply(aggData_, val);
+    DCHECK(aggFunc_.hasValue());
```
Comment (Shylock-Hg) [inline src/common/expression/AggregateExpression.cpp:50]: How about invalid agg function name?

Comment (czpmango) [reply src/common/expression/AggregateExpression.cpp:50]: Checked in parser or validator. Invalid function name is not expected here.

Comment (Shylock-Hg) [reply src/common/expression/AggregateExpression.cpp:50]: Do we have the case in tck?

Comment (czpmango) [reply src/common/expression/AggregateExpression.cpp:50]: Do you mean this case:
```
(czp@nebula) [nba]> go from "Tim Duncan" over like yield $$.player.age as age | group by $-.age yield $-.age,count($-.age) as count
+--------+-------+
| $-.age | count |
+--------+-------+
| 38     | 2     |
+--------+-------+
| 41     | 1     |
+--------+-------+
| 36     | 1     |
+--------+-------+
Got 3 rows (time spent 387690/395754 us)

Wed, 24 Mar 2021 13:59:12 CST

(czp@nebula) [nba]> go from "Tim Duncan" over like yield $$.player.age as age | group by $-.age yield $-.age,counte($-.age) as count
[ERROR (-7)]: SyntaxError: Unknown function  near `counte'
```

Comment (Shylock-Hg) [reply src/common/expression/AggregateExpression.cpp:50]: Yes, it's ok for me

**Review by Shylock-Hg — APPROVED** [review]

File: src/common/expression/AggregateExpression.h [hunk 2]
```diff
@@ -78,15 +84,16 @@ class AggregateExpression final : public Expression {
...
     void writeTo(Encoder& encoder) const override;
     void resetFrom(Decoder& decoder) override;
 
     std::unique_ptr<std::string>    name_;
     std::unique_ptr<Expression>     arg_;
     bool                            distinct_{false};
     AggData*                        aggData_{nullptr};
+
+    // runtime cache for aggregate function lambda
+    folly::Optional<AggFunctionManager::AggFunction> aggFunc_;
```
Comment (Shylock-Hg) [inline src/common/expression/AggregateExpression.h:96]: If the invalid function name is disabled in parser, I think you could remove the optional.

**Review by yixinglu — COMMENTED** [review]
Lookup my comments in #407 , my suggestion is that never to check whether this function is nullable in `eval` function. It's in the hot path when evaluate expression. Please ensure the function existence when construct the aggr func call expression.

File: src/common/expression/AggregateExpression.h (hunk 2 above)
Comment (czpmango) [reply src/common/expression/AggregateExpression.h:96]: ACK.

**Comment by czpmango** [comment]
> Lookup my comments in #407 , my suggestion is that never to check whether this function is nullable in `eval` function. It's in the hot path when evaluate expression. Please ensure the function existence when construct the aggr func call expression.

ACK.

**Review by jievince — APPROVED** [review]

File: src/common/expression/AggregateExpression.h [hunk 3]
```diff
@@ -23,6 +23,12 @@ class AggregateExpression final : public Expression {
         arg_.reset(arg);
         name_.reset(name);
         distinct_ = distinct;
+        if (name_) {
+            auto aggFuncResult = AggFunctionManager::get(*name_);
+            if (aggFuncResult.ok()) {
+                aggFunc_ = std::move(aggFuncResult).value();
+            }
+        }
```
Comment (yixinglu) [inline src/common/expression/AggregateExpression.h:31]: statici StatusOr<AggregateExpression*> create(name,xxx) {
  fn = AggFmgr.get(name);
  if (!fn) return Status::Error();
  return new AggregateFnExpr(fn);
}

Comment (czpmango) [reply src/common/expression/AggregateExpression.h:31]: ACK.

Comment (czpmango) [reply src/common/expression/AggregateExpression.h:31]: As we discussed offline, I tried to modify the interface above but got stuck by the memory management conflict of the parser, so the interface here is still maintained. Modify it if we unify the `Expression::create` interface later.

**Review by yixinglu — DISMISSED** [review]

**Review by nevermore3 — DISMISSED** [review]

**Review by yixinglu — APPROVED** [review]

**Review by Aiee — APPROVED** [review]

**Pull request merged**